Поймать и приручить слона

Ivan-Pelipenko opened the pull request:
---
:mortar_board: [Поймать и приручить слона](https://up.htmlacademy.ru/nodejs-2/4/user/2176283/tasks/8)

Ivan-Pelipenko committed:
4.1. Добавит новый проект blog.
Сгенерируем основу для следующего сервиса «Blog». Этот сервис отвечает
за управление контентом. Он предоставляет API для создания, удаления,
редактирования контента, а также управление комментариями.

Для создания новой заготовки воспользуемся командой:

```
nx generate @nrwl/nest:app blog
```

Ivan-Pelipenko committed:
4.2. Добавит docker-compose.dev.yml для Docker.
СУБД PostgreSQL удобней установить в Docker. Подготовка
compose-файла для PostgreSQL не сильно отличается от аналогичной
процедуры для Mongo.

Обратим внимание на порты и директорию, в которой будет
создана база данных. Мы прокидываем порт 5435 (порт по умолчанию для
PostgreSQL), а для хранения данных воспользуемся томами.

Помимо PostgreSQL потребуется инструмент для администрирования базы.
Мы воспользуемся PGAdmin. Его установку произведём также с помощью Docker.
Для этого добавим отдельный сервис `pgadmin` в `docker-compose`.

Обратите внимание: все значения для переменных окружения мы передаём
прямо из самого `docker-compose`. При необходимости можно
воспользоваться директивой `env_file` и указать путь к файлу
с установленными переменными окружениями.

Для сервиса с PostgreSQL добавим секцию `healthcheck`. В ней
опишем проверку работоспособности сервиса, который запускаем
в контейнере.

Например, мы запускаем PostgreSQL, поэтому разумно настроить
периодическую проверку доступности БД. Если служба БД перестаёт
отвечать, значит что-то пошло не так. Этот факт отразится в логах
и контейнер сможет перезапуститься.

Для проверки состояния PostgreSQL мы воспользуемся утилитой `pg_isready`.
Утилита предназначена для проверки соединения с PostgreSQL. Смысл
работы прост: происходит попытка установить соединение с сервером СУБД.
Если соединиться не вышло, утилита вернёт соответствующий код ошибки.

Обратите внимание, для работы `pg_isready` требуется имя пользователя и
пароль. Мы передаём те же данные, которые указали в секции `environment`.

Ivan-Pelipenko committed:
4.3. Установит Prisma ORM.
Для взаимодействия с PostgreSQL воспользуемся ORM. Он упростит
взаимодействие с базой данных и в большинстве случае избавит от
необходимости подготавливать запросы на языке SQL. Вместо SQL мы сможем
работать с данными из БД как с объектами. Однако, знания SQL по-прежнему
остаются необходимыми, так как в некоторых случаях, подготовка запроса
на SQL может упростить решение задачи.

Ivan-Pelipenko committed:
4.4. Добавит новую библиотеку models/blog-models.

Ivan-Pelipenko committed:
4.5. Инициализирует Prisma.
Перед тем как начать применять Prisma ORM, необходимо создать заготовку
для описания схемы базы данных.

Для этого необходимо инициализировать Prisma для проекта. Для
инициализации выполним команду (из директории библиотеки):

```
npx prisma init
```

Команда создаст директорию `prisma`, а в ней файл `schema.prisma`.
В этом файле необходимо описать схему, по которой Prisma сможет
сгенерировать схему базу данных.

Ivan-Pelipenko committed:
4.6. Добавим описание моделям Post.

Добавим в файл `schema.prisma` описание моделей — `Post`. На
основании этого описания `Prisma` сможет сгенерировать соответствующие
таблицы в схеме базы данных.

При описании моделей мы перечисляем все поля, перечисленные в
техническом задании. Prisma предоставляет все необходимые типы для
определения полей, а также дополнительные модификаторы. Например,
модификатор `@id` для определения поля с идентификатором. Вместе с
ним мы применяем аннотацию `@default`. В ней можно определить значение
по умолчанию для поля. Аргументом в `@default` передаётся вызов функции
`autoincrement()`. Она позволяет создать последовательность целых чисел,
которые используются в качестве идентификатора. В PostgreSQL она
соответствует типу `SERIAL`.

После описания моделей, мы можем их проверить на корректность,
или проще сказать: пролинтерить. Для этого выполним команду:

```
npx prisma format --schema ./libs/models/blog-model/prisma/schema.prisma
```

В аргументе `--schema` передадим путь к созданной схеме. При обнаружении
синтаксических ошибок, информация будет выведена в консоль.

Следующим шагом попробуем сформировать схему базы данных на основании
Prisma схема. Для этого вновь придётся обратиться к командной строке:

```
npx prisma migrate dev --name "added post model" --schema ./libs/models/blog-models/prisma/schema.prisma --skip-generate
```

Команда `migrate dev` используется только на этапе разработки. Её нельзя
применять в продуктовом режиме.

Аргументы:

* `--name` — название миграции
* `--schema` — путь к схеме
* `--skip-generate` — пропустить формирование клиента.

Добавим модификатор `@@Map()` в описание
схемы и передадим произвольные имена для таблиц. Если требуется дать другое имя
для поля на уровне БД воспользуйтесь модификатором `@map`.

NaumovEvgeniy commented on project/apps/blog/project.json:
Не реализованы команды из пункта 4 и 6 домашки

> 4. Добавьте project.json сервиса, который использует команды Prisma ORM для быстрого выполнения операций с базой. Например: db-validate (для проверки схемы), db-generate (для генерации Prisma Client), db-reset (для удаления и очистки БД), db-migrate (для запуска миграции). В качестве executor используйте nx:run-commands.

> 6. Добавьте в project.json комаду db-fill. Она также использует nx:run-commands. Выполнение команды должно приводить к наполнению базы данных.


К сожалению, остальной функционал без этих правок не смогу проверить. 

Если есть вопросы, пиши, спрашивай. Можно голосом обсудить, потому что тема с pgsql сложная и тут есть тонкости и подводные камни

Ivan-Pelipenko committed:
4.10. Добавит описание модели Comment
Подготовим модель `Comment`. Она определяет сущность
«Комментарий» в блоге.

Ivan-Pelipenko committed:
4.11. Добавит команду db:lint для сервиса blog.
Взаимодействовать с Prisma CLI придётся часто. Поэтому удобней подготовить
отдельные команды для быстрого запуска. Обычно такие вещи описываем
в `package.json` внутри секции `scripts`. Поскольку используем
монорепозиторий, опишем в `project.json` отдельного проекта.

При описании команд мы можем указать произвольную функцию исполнитель
(`executor`), которая будет вызываться. Мы указываем `nx:run-commands`.
Это специальная функция от `nx`, мы можем её использовать для подготовки
своих команд.

В качестве первой команды создадим `db:lint`. Она выполняет вызов
`prisma` и передаёт в качестве выполнения команду `validate`. Она
выполняет проверку схемы (`schema.prisma`).

Для запуска добавленной команды необходимо выполнить:

```
nx run blog:db:lint
```

Ivan-Pelipenko committed:
4.13. Добавит команды db:migrate и db:reset.

Ivan-Pelipenko committed:
4.14. Добавит команду db:generate и установит пакет @prisma/client.
Prisma Client — дополнительный пакет, он предоставляет клиент для
базы данных. Через него мы сможем обращаться к базе данных и выполнять
различные операции. Prisma Client создаётся на основе определения
модулей Prisma. Автоматически будут подготовлены все необходимые типы.

Для быстрой подготовки обновлённого клиента нам потребуется ещё
одна команда для проекта: `db:generate`. Она запустит генерацию клиента для Prisma.

Перед запуском команды необходимо внести изменения в `prisma.schema`.
Пока Prisma не учитывает многие нюансы монорепозиториев, поэтому нужно самостоятельно указать путь, по которому должен генерироваться клиент.

Ivan-Pelipenko committed:
4.15. Добавит сидирование (наполнение) базы данных.
Сидирование базы данных - это процесс начальной инициализации базы данных, при которой ей предоставляются первоначальные данные.

В общем случае, сидирование базы данных происходит в том случае, если база данных пуста или содержит только небольшое количество данных. В процессе сидирования базы данных можно заполнить ее данными, чтобы в дальнейшем можно было работать с полноценной базой.

Часто сидирование базы данных происходит с использованием скриптов, которые выполняются при создании базы данных. Эти скрипты могут содержать SQL-запросы для создания таблиц, заполнения их начальными данными и создания индексов.

Для сидирования базы данных воспользуемся Prisma ORM.

Ivan-Pelipenko committed:
4.16. Добавит команду db:seed.
Для быстрого запуска добавим отдельную команду `db:seed`. Она выполнит запуск сценария `seed.ts`. Таким образом, для начального наполнения базы данных у нас появится команда:

```
nx run blog:db:seed
```

Ivan-Pelipenko committed:
4.17. Добавит интерфейс Entity.
Напомним, в контексте работы с базами данных, сущность (Entity) - это объект который представляет определенный тип данных и хранится в базе данных. У всех Entity есть общее поведение. Определим его в интерфейсе Entity.

Ivan-Pelipenko committed:
4.18. Добавит интерфейсы для сущностей блога.
Создадим интерфейсы `Post`, `Comment`, `Like`. Все интерфейсы
создаём в общей библиотеке «app-types».

Ivan-Pelipenko committed:
4.19. Добавит модуль PrismaModule.
Создадим отдельный модуль и реализуем в нём взаимодействие с
Prisma Client. Для этого потребуется определить отдельный сервис и
класс с модулем.

Ivan-Pelipenko committed:
4.20. Добавит сущность BlogCommentEntity.

Ivan-Pelipenko committed:
4.21. Добавит BlogCommentRepository.
Подготовим отдельный репозиторий для управления комментариями.

Ivan-Pelipenko committed:
4.22. Добавит сервис BlogCommentService и dto.

Ivan-Pelipenko committed:
4.23. Добавит контроллер BlogCommentController.
Реализуем обработчики маршрутов для работы с блогом.

Ivan-Pelipenko committed:
4.24. Добавит модуль BlogCategoryModule.

Ivan-Pelipenko committed:
4.25. Добавит заготовки запросов к контроллеру BlogCommentController.

Подключит PrismaModule и BlogCommentModule в AppModule.
